aiobotocore/httpxsession.py

```diff
@@ -0,0 +1,305 @@
+from botocore.httpsession import (
+    MAX_POOL_CONNECTIONS,
+    EndpointConnectionError,
+    HTTPClientError,
+    ReadTimeoutError,
+    create_urllib3_context,
+    ensure_boolean,
+    get_cert_path,
+    logger,
+)
+from multidict import CIMultiDict
+
+import aiobotocore.awsrequest
+from aiobotocore._endpoint_helpers import _text
+
+# TODO: resolve future annotations thing
+
+
+try:
+    import httpx
+except ImportError:
+    httpx = None
+
+if TYPE_CHECKING:
+    from ssl import SSLContext
+
+
+class HttpxSession:
+    def __init__(
+        self,
+        verify: bool = True,
+        proxies: dict[str, str] | None = None,  # {scheme: url}
+        timeout: float | list[float] | tuple[float, float] | None = None,
+        max_pool_connections: int = MAX_POOL_CONNECTIONS,
+        socket_options: list[Any] | None = None,
+        client_cert: str | tuple[str, str] | None = None,
+        proxies_config: dict[str, str] | None = None,
+        connector_args: dict[str, Any] | None = None,
+    ):
+        if httpx is None:  # pragma: no cover
+            raise RuntimeError(
+                "Using HttpxSession requires httpx to be installed"
+            )
+        if proxies or proxies_config:
+            raise NotImplementedError(
+                "Proxy support not implemented with httpx as backend."
+            )
+
+        # TODO: handle socket_options
+        self._session: httpx.AsyncClient | None = None
+        conn_timeout: float | None
+        read_timeout: float | None
+
+        if isinstance(timeout, (list, tuple)):
+            conn_timeout, read_timeout = timeout
+        else:
+            conn_timeout = read_timeout = timeout
+        # must specify a default or set all four parameters explicitly
+        # 5 is httpx default default
+        self._timeout = httpx.Timeout(
+            5, connect=conn_timeout, read=read_timeout
+        )
+
+        self._cert_file = None
+        self._key_file = None
+        if isinstance(client_cert, str):
+            self._cert_file = client_cert
+        elif isinstance(client_cert, tuple):
+            self._cert_file, self._key_file = client_cert
+        elif client_cert is not None:
+            raise TypeError(f'{client_cert} must be str or tuple[str,str]')
+
+        # previous logic was: if no connector args, specify keepalive_expiry=12
+        # if any connector args, don't specify keepalive_expiry.
+        # That seems .. weird to me? I'd expect "specify keepalive_expiry if user doesn't"
+        # but keeping logic the same for now.
+        if connector_args is None:
+            # aiohttp default was 30
+            # AWS has a 20 second idle timeout:
+            #   https://web.archive.org/web/20150926192339/https://forums.aws.amazon.com/message.jspa?messageID=215367
+            # "httpx default timeout is 5s so set something reasonable here"
+            self._connector_args: dict[str, Any] = {'keepalive_timeout': 12}
```

Comment:
may want to make this an official param as well

Reply:
it already is, cleaned up the logic so it's easier to follow and so `12` isn't specified in three different files for the same thing